pkg/filter/filter.go

```diff
@@ -108,24 +119,44 @@ func JoinResourceTypeMatchers(ms ...ResourceTypeMatcher) ResourceTypeMatcher {
 
 // ResourceRequirement allows specifying a resource requirement by type and/or name
 type ResourceRequirement struct {
+	// Provides the Name of the resource object
 	v1.LocalObjectReference `json:",inline,omitempty"`
+	// Provides the Group, Version, and Resource values (GVR)
 	ResourceTypeRequirement `json:",inline,omitempty"`
+	// Specifies a set of label requirements to be used as filters for matches
+	metav1.LabelSelector `json:",inline,omitempty"`
 }
 
-// Matches returns true if the specified resource name/GVR matches the requirement
-func (r ResourceRequirement) Matches(name string, gvr schema.GroupVersionResource) bool {
-	// If the requirement does not specify a resource name - only check the
-	// ResourceTypeRequirement i.e. GVR match
-	if r.LocalObjectReference.Name == "" {
-		return r.ResourceTypeRequirement.Matches(gvr)
+// Matches returns true if the specified resource name/GVR/labels matches the requirement
+func (r ResourceRequirement) Matches(name string, gvr schema.GroupVersionResource, resourceLabels map[string]string) bool {
+	// If ResourceRequirement Name/GVR value of empty string matches on any specified in parameters.
+	// If ResourceRequirement Name/GVR value specified, exact match on parameter is required.
+	if !(matches(r.Name, name) && r.ResourceTypeRequirement.Matches(gvr)) {
+		return false
 	}
-	return matches(r.Name, name) && r.ResourceTypeRequirement.Matches(gvr)
+	if len(resourceLabels) == 0 && len(r.MatchExpressions) == 0 && len(r.MatchLabels) == 0 {
+		// If there are no resourceLabels and no LabelSelector we return match of GVR from above.
+		return true
+	}
+	if len(resourceLabels) == 0 && (len(r.MatchExpressions) != 0 || len(r.MatchLabels) != 0) {
```

Comment:
This does not handle the case where the requirement is an exclusion label.  Such as `r.MatchExpressions{{key: foo,  operator: NotIn, values: [bar]}}`.

Reply:
This is true. Do we have to assume that if the caller does not supply resourceLabels, the MatcheExpressions should be ignored? If a resource has no labels and the user wants to include labels NotIn [list], then we would want to return a match true given that the resource has no labels "in" that list. Let me experiment and see what I can come up with.

Reply:
Removed that check and added test. You're right. It's not needed.